### ISSUE TITLE
[BUG] LPC176x (SKR 1.4 Turbo) boards and MAX31855 not working on E1 empty Driver Socket in Software SPI mode

### ISSUE THREAD

**Issue by GadgetAngel** [MarlinFirmware/Marlin#20418]
### Bug Description

Since the SKR V1.3 board and the SKR V1.4 and SKR V1.4 Turbo boards do not have a lot of extra free pins, I want to use the free pins on the empty E1 Driver socket to run a K-Type Thermocouple temperature sensor. I do this on the SKR PRO and GTR boards without an issue.

I have an SKR V1.4 Turbo board I am trying to hook up a K-Type thermocouple with a MAX31855 board from Adafruit.  When I use the empty E1 driver socket pins (E1_DIR_PIN , E1_STEP_PIN and E1_PDN_PIN) all I get is MIN TEMP error.  So I double check my work and I still get MIN TEMP error.  I went and bought myself a USB logic analyzer from Saleae (PRO 8 channel).  I will be supplying the capture SPI bus data with this bug report.  

My SKR V1.4 Turbo board is using 3 UART drivers TMC2226 in X, Y, and Z.  The E0 Driver is in SPI mode and is a TMC2130.  I ensure I comment out the E1 pins defined in the pins_BTT_SKR_V1_4.h file and in pins_BTT_SKR_common.h file.  I comment out the CS_PIN for all drivers not using SPI mode.  I comment out ?_SERIAL_TX_PIN and ?_SERIAL_RX_PIN for all drivers not in UART mode:
pins_SKR_V1_4.h file:
```
//
// Using UART mode on X, Y and Z BUT using SPI mode on E0, E1 NOT being used
//

#define X_STEP_PIN                         P2_02
#define X_DIR_PIN                          P2_06
#define X_ENABLE_PIN                       P2_01
//#ifndef X_CS_PIN
  //#define X_CS_PIN                         P1_10
//#endif

#define Y_STEP_PIN                         P0_19
#define Y_DIR_PIN                          P0_20
#define Y_ENABLE_PIN                       P2_08
//#ifndef Y_CS_PIN
  //#define Y_CS_PIN                         P1_09
//#endif

#define Z_STEP_PIN                         P0_22
#define Z_DIR_PIN                          P2_11
#define Z_ENABLE_PIN                       P0_21
//#ifndef Z_CS_PIN
  //#define Z_CS_PIN                         P1_08
//#endif

#define E0_STEP_PIN                        P2_13
#define E0_DIR_PIN                         P0_11
#define E0_ENABLE_PIN                      P2_12
#ifndef E0_CS_PIN
  #define E0_CS_PIN                        P1_04
#endif

/*
#define E1_STEP_PIN                        P1_15
#define E1_DIR_PIN                         P1_14
#define E1_ENABLE_PIN                      P1_16
//#ifndef E1_CS_PIN
  //#define E1_CS_PIN                        P1_01
//#endif
*/

//#define TEMP_1_PIN                      P0_23_A0  // A0 (T0) - (67) - TEMP_1_PIN
#define TEMP_BED_PIN                    P0_25_A2  // A2 (T2) - (69) - TEMP_BED_PIN

//
// Software SPI pins for TMC2130 stepper drivers
//
#if ENABLED(TMC_USE_SW_SPI)
  #ifndef TMC_SW_MOSI
    #define TMC_SW_MOSI                    P1_17
  #endif
  #ifndef TMC_SW_MISO
    #define TMC_SW_MISO                    P0_05
  #endif
  #ifndef TMC_SW_SCK
    #define TMC_SW_SCK                     P0_04
  #endif
#endif

#if HAS_TMC_UART
  /**
   * TMC2208/TMC2209 stepper drivers
   *
   * Hardware serial communication ports.
   * If undefined software serial is used according to the pins below
   */
  //#define X_HARDWARE_SERIAL  Serial1
  //#define X2_HARDWARE_SERIAL Serial1
  //#define Y_HARDWARE_SERIAL  Serial1
  //#define Y2_HARDWARE_SERIAL Serial1
  //#define Z_HARDWARE_SERIAL  Serial1
  //#define Z2_HARDWARE_SERIAL Serial1
  //#define E0_HARDWARE_SERIAL Serial1
  //#define E1_HARDWARE_SERIAL Serial1
  //#define E2_HARDWARE_SERIAL Serial1
  //#define E3_HARDWARE_SERIAL Serial1
  //#define E4_HARDWARE_SERIAL Serial1

  //
  // Software serial
  //
  #define X_SERIAL_TX_PIN                  P1_10
  #define X_SERIAL_RX_PIN                  P1_10

  #define Y_SERIAL_TX_PIN                  P1_09
  #define Y_SERIAL_RX_PIN                  P1_09

  #define Z_SERIAL_TX_PIN                  P1_08
  #define Z_SERIAL_RX_PIN                  P1_08

  //#define E0_SERIAL_TX_PIN                 P1_04
  //#define E0_SERIAL_RX_PIN                 P1_04

  //#define E1_SERIAL_TX_PIN                 P1_01
  //#define E1_SERIAL_RX_PIN                 P1_01

  // Reduce baud rate to improve software serial reliability
  #define TMC_BAUD_RATE                    19200
#endif
```
pins_BTT_SKR_common.h file:
```
#pragma once

#ifdef SKR_HAS_LPC1769
  #if NOT_TARGET(MCU_LPC1769)
    #error "Oops! Make sure you have the LPC1769 environment selected in your IDE."
  #endif
#elif NOT_TARGET(MCU_LPC1768)
  #error "Oops! Make sure you have the LPC1768 environment selected in your IDE."
#endif

// If you have the Big tree tech driver expansion module, enable HAS_BTT_EXP_MOT
// https://github.com/bigtreetech/BTT-Expansion-module/tree/master/BTT%20EXP-MOT
//#define HAS_BTT_EXP_MOT 1

#if BOTH(HAS_WIRED_LCD, HAS_BTT_EXP_MOT)
  #if EITHER(CR10_STOCKDISPLAY, ENDER2_STOCKDISPLAY)
    #define EXP_MOT_USE_EXP2_ONLY
  #else
    #error "Having a LCD that uses both EXP1/EXP2 and a expansion motor module on EXP1/EXP2 is not possible."
  #endif
#endif

// Ignore temp readings during development.
//#define BOGUS_TEMPERATURE_GRACE_PERIOD    2000

//
// Steppers
//

/* E1 not USED
#ifndef E1_STEP_PIN
  #define E1_STEP_PIN                      P0_01
#endif
#ifndef E1_DIR_PIN
  #define E1_DIR_PIN                       P0_00
#endif
#ifndef E1_ENABLE_PIN
  #define E1_ENABLE_PIN                    P0_10
#endif
*/

//
// Temperature Sensors
//  3.3V max when defined as an analog input
//
#ifndef TEMP_0_PIN
  #define TEMP_0_PIN                    P1_24  // A1 (T1) - (68) - TEMP_0_PIN
#endif
#ifndef TEMP_1_PIN
  #define TEMP_1_PIN                    P0_25_A2  // A2 (T2) - (69) - TEMP_1_PIN
#endif
#ifndef TEMP_BED_PIN
  #define TEMP_BED_PIN                  P0_23_A0  // A0 (T0) - (67) - TEMP_BED_PIN
#endif
```

Note: the following convention I will use when discussing this issue:
```
E1_PDN_PIN is the E1_CS_PIN which is equal to E1_SERIAL_TX_PIN and E1_SERIAL_RX_PIN which is P1_01
E1_STEP_PIN is P1_15
E1_DIR_PIN is P1_14
```
When capturing the SPI signals, I always start the capture software (Logic 2 Beta) first.  I then run the boot process and then take data for a while and then shut off power to the SKR V1.4 Turbo board before stopping the software that captures the SPI lines. This way you get a full picture of what is going on with the SPI lines.

I preformed 4 tests runs and captured the results:

Test number 1: 
I am using P1_24 which is the Neopixel connector's signal pin as TEMP_0_PIN in pins_BTT_SKR_common.h file (MAX6675_SS_PIN is equal to TEMP_0_PIN).
Using P0_03 which is the TFT connector's RX0 signal pin as MAX6675_DO_PIN
Using P1_01 which is the E1_PDN_PIN as MAX6675_SCK_PIN. The "I" Jumper from "SPIM" is in place on the internal E1 driver socket pins, where "S" is the top row "P" is the second row and "I" is the third row and "M" is the bottom row.  The "I" Jumper is placed on the left two pins on the third row of pins.

So Test number 1 Setup:
```
MAX6675_SS_PIN     P1_24
MAX6675_DO_PIN     P0_03
MAX6675_SCK_PIN    P1_01   //E1_PDN_PIN with "I" jumper in place
```
Test number 1 Result:
```
With this configuration all work BUT I am only using one of the E1 socket PINS! I want it to work with all three E1 PINS.
```

Test number 2:
Interesting enough if I move the P1_01 to act as the MAX6675_DO_PIN and use P0_02 to act as the MAX6675_SCK_PIN, and keep P1_24 as MAX6675_SS_PIN, I get the same **MIN TEMP error**! P0_02 is the TFT connector's TX0 signal pin.  

Test number 2 Setup:
```
MAX6675_SS_PIN     P1_24
MAX6675_DO_PIN     P1_01   //E1_PDN_PIN with "I" jumper in place
MAX6675_SCK_PIN    P0_02
```
Test number 2 Results:
```
What is really interesting is the what the captured SPI data shows: You can see ONE temperature sample (32 bits of data) was asked for and received by the SKR V1.4 Turbo board and then the CS line stays high and the MISO line stays low and the SCK line stays low after the first 32 bits are read.  It looks like something in Marlin is not allowing the lines to be used.  
```

So I tried another test (Test number 3) this time I will use P1_15 (E_STEP_PIN).  I use it like I did in Test number 1 as the MAX6675_SCK_PIN.

Test number 3 Setup:
```
MAX6675_SS_PIN     P1_24
MAX6675_DO_PIN     P0_03
MAX6675_SCK_PIN    P1_15
```
Test number 3 Results:
```
With this configuration all work BUT I am only using one of the E1 socket PINS! I want it to work with all three E1 PINS.
```
So I move the P1_15 to act as MAX6675_DO_PIN for Test number 4.

Test number 4 Setup:
```
MAX6675_SS_PIN     P1_24
MAX6675_DO_PIN     P1_15
MAX6675_SCK_PIN    P0_02
```
Test number 4 Results, I get a **MIN TEMP error**!:
```
What is really interesting is the what the captured SPI data shows: You can see ONE temperature sample (32 bits of data) was asked for and received by the SKR V1.4 Turbo board and then the CS line stays high and the MISO line stays low and the SCK line stays low after the first 32 bits are read.  It looks like something in Marlin is not allowing the lines to be used. 
```

So now I have a repeatable set of test that shows that the SKR V1.4 Turbo should be able to use the E1 PINS (P1_01, P1_14, and P1_15) but something is stopping it from happening. 

For my own sake of mind, I went ahead and used the **MAX31855 library call from Adafruit MAX31855 library** for Test number 5.  

Test number 5 Setup:
```
MAX6675_SS_PIN     P1_14
MAX6675_DO_PIN     P1_15
MAX6675_SCK_PIN    P1_01
```
Test number 5 Result:
```
ONLY with the MAX31855 Adafruit Library call could I get ALL the E1 PINS to work in Software SPI mode on the SKR V1.3 or SKR V1.4 or SKR V1.4 Turbo boards.  THIS WORKS!
I will be submitting this as a PR but my PR will only force the use of the Adafruit MAX31855 library call for the TARGET_LPC1768 boards.  

Why this is not working without the Adafruit library call?
```

### Configuration Files

[Configuration files for MAX31855 bug.zip](https://github.com/MarlinFirmware/Marlin/files/5669009/Configuration.files.for.MAX31855.bug.zip)

### Steps to Reproduce

1. Use an SKR V1.3 or SKR V1.4 or SKR V1.4 Turbo board, use 3 UART stepper motor drivers in X, Y and Z. Use one SPI stepper motor driver in E0
2. comment out the E1 pins in pins_BTT_SKR_V1.x.h file and pins_BTT_SKR_common.h file.  Ensure you comment out the ??_CS_PIN for the driver sockets that are in UART mode and comment out the ??_SERIAL_TX_PIN and ??_SERIAL_RX_PIN for the E0 and E1 driver sockets since they are NOT in UART mode
3. On E1 driver socket that is empty place a jumper on the internal jumpers in the third row down and on the left two pins just like you would if you had a driver in UART mode.  This jumper will connect the MCU pin to the M3 pin of the E1 driver socket.
4. in the pins_BTT_SKR_common.h file define the following:
```
TEMP_0_PIN         P1_14
MAX6675_SS_PIN     TEMP_0_PIN
MAX6675_DO_PIN     P1_15
MAX6675_SCK_PIN    P1_01
```
in configuration.h:
```
#define SERIAL_PORT -1
//#define SERIAL_PORT_2 0
#define BAUDRATE 115200
#define EXTRUDERS 1
#define TEMP_SENSOR_0 -3
#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
```
in configuration_adv.h:
```
#define SHOW_TEMP_ADC_VALUES
#define TMC_USE_SW_SPI
#define MONITOR_DRIVER_STATUS
#define TMC_DEBUG
```

**Expected behavior:**

I want to use the E1 empty driver socket pins to run a MAX31855 board and K-type thermocouple.  So I want to use:
```
MAX6675_SS_PIN     P1_14
MAX6675_DO_PIN     P1_15
MAX6675_SCK_PIN    P1_01
and set TEMP_SENSOR_0 to -3
```
I expect my printer to use the temperature readings from the thermocouple and NOT to get and MIN TEMP error on boot up.

**Actual behavior:**

I get a MIN TEMP error! on bootup of the SKR V1.4 Turbo board.

#### Additional Information

Here are the SPI capture files I was describing above.  Did you all know that the **first clock pulse** and the **second clock pulse** produced by the LPC176x board **is less than 100 ns apart**.  The MAX31855 datasheet and other SPI data sheets all say that a minimum of 100 ns on the negative clock pulse width is needed. You will see this in the SPI capture data sets.

You can download the Saleae Software (2.3.15) from here and use it to open up these files to look at them for yourself: https://ideas.saleae.com/f/changelog/

Documents on how to use the Saleae Software can be found here: https://support.saleae.com/getting-started

Test number 1 SPI capture: 
[Test number 1 SPI capture data set.zip](https://github.com/MarlinFirmware/Marlin/files/5669116/P1_01.acting.as.SCK.line.with.P1_24.as.CS.line.and.P0_03.as.DO_PIN.appears.to.be.OK.zip)

Test number 2 SPI capture: THIS IS where you see a problem (ONLY ONE SAMPLE is allowed through!): 
[Test number 2 SPI capture data set.zip](https://github.com/MarlinFirmware/Marlin/files/5669128/P1_01.acting.as.D0_PIN.with.P1_24.as.CS.and.P0_02.as.SCK.ONE.reading.then.0.zip)

Test number 3 SPI capture: 
[Test number 3 SPI capture data set.zip](https://github.com/MarlinFirmware/Marlin/files/5669133/P1_15.acting.as.SCK_PIN.with.P1_24.as.CS.and.P0_03.as.DO_PIN.everything.appears.OK.zip)

Test number 4 SPI capture: THIS IS where you see a problem (ONLY ONE SAMPLE is allowed through!)  Now that this happened a second time with a different pin the problem is repeatable.  
[Test number 4 SPI capture data set.zip](https://github.com/MarlinFirmware/Marlin/files/5669146/P1_15.acting.as.DO_PIN.with.P1_24.acting.as.CS.and.P0_02.acting.as.SCK_PIN.ONLY.one.sample.then.zero.values.zip)

Here is another configuration file (for this issue, I call issue#5):
[Configurations_Issue#5_MAX31855 will not work on E1 socket with SKR V1.4 Turbo.zip](https://github.com/MarlinFirmware/Marlin/files/5682459/Configurations_Issue.5_MAX31855.will.not.work.on.E1.socket.with.SKR.V1.4.Turbo.zip)

Here is a link to a video showing Issue#5: https://drive.google.com/file/d/1gfes4-4VozdZxBdNdRCTonhXl6iNBOfZ/view?usp=sharing

**Comment by rhapsodyv** [comment]
It's not the same as #20359, but on different pins? Looks like the problem is SW with MAX sensors on any pin set...

**Comment by GadgetAngel** [comment]
@rhapsodyv BUG report #20359 is about the MAX6675 dropping temperature readings down to 0 for no known reason.  

This BUG report is the inability to use the empty E1 driver socket pins to run a MAX31855 board. 

The next report you will see is #20429 which is the inability to use Hardware SPI with RX0 pin on the TFT connector.  

It feels like the LPC176x boards are running on the edge of getting data with SPI.  If you look at the SPI captures I sent with this bug report you will see that the clock pulses are not the recommended 100ns pulse width.  

I will be submitting a PR that will fix this bug report and #20429 bug report.   They both are fixed by using the Adafruit MAX31855 library for the LPC1768 board.  The Adafruit MAX31855 library spaces the clock pulses to 1ms instead of less than 100ns.  I am testing my PR to ensure that it does not cause more issues to come up.

**Comment by rhapsodyv** [comment]
You reported 4 issues about MAX sensors (31855 or 6675), all on LPC.... I really think that they share the same cause.

But as I soon I receive mine, I will take a deep look. I think they will arrive only after holidays.

**Comment by GadgetAngel** [comment]
> This issue has had no activity in the last 30 days. Please add a reply if you want to keep this issue active, otherwise it will be automatically closed within 7 days.

This issue is still under review because the pull request that fixes this issue is still going through the review process.

**Comment by GadgetAngel** [comment]
This issue is resolved because the PR #20447 has now been merged in bugfix branch
